actions.js

```diff
@@ -0,0 +1,234 @@
+	{ label: 'NET/USB', id: 'NET/USB' },
+	{ label: 'XM', id: 'XM' },
+	{ label: '8K', id: '8K' },
+	{ label: 'IPOD', id: 'IPOD' },
+	{ label: 'Same as zone 1', id: 'SOURCE' },
+]
+
+module.exports = function (self) {
+	CHOICES_VOLUME = []
+
+	for (i = 98; i >= 1; i--) {
+		db = i - 80
+		v = String(i).padStart(2, '0')
+		CHOICES_VOLUME.push({ label: db + 'dB', id: '' + v + '' })
+	}
+
+	CHOICES_VOLUME.push({ label: 'Mute', id: '00' })
+
+	self.setActionDefinitions({
+		power: {
+			name: 'Power control',
+			options: [
+				{
+					type: 'dropdown',
+					label: 'on/standby',
+					id: 'power',
+					default: 'PWON\r',
+					choices: [
+						{ label: 'Power on', id: 'PWON' },
+						{ label: 'standby', id: 'PWSTANDBY' },
+						{ label: 'Zone2 on', id: 'Z2ON' },
+						{ label: 'Zone2 off', id: 'Z2OFF' },
+						{ label: 'Zone3 on', id: 'Z3ON' },
+						{ label: 'Zone3 off', id: 'Z3OFF' },
+					],
+				},
+			],
+			callback: async (event) => {
+				var cmd = event.options.power + '\r'
+				console.log('command sent to denon ->', cmd)
+				self.socket.send(cmd)
+			},
+		},
+		mute: {
+			name: 'Mute controls',
+			options: [
+				{
+					type: 'dropdown',
+					label: 'zone/on/off',
+					id: 'mute',
+					default: 'MUON\r',
+					choices: [
+						{ label: 'Main zone mute on', id: 'MUON\r' },
+						{ label: 'Main zone mute off', id: 'MUOFF\r' },
+						{ label: 'Zone 2 mute on', id: 'Z2MUON\r' },
+						{ label: 'Zone 2 mute off', id: 'Z2MUOFF\r' },
+						{ label: 'Zone 3 mute on', id: 'Z3MUON\r' },
+						{ label: 'Zone 3 mute off', id: 'Z3MUOFF\r' },
+					],
+				},
+			],
+			callback: async (event) => {
+				var cmd = event.options.mute
+				console.log('command sent to denon ->', cmd)
+				self.socket.send(cmd)
+			},
+		},
+		volume: {
+			name: 'Volume',
+			options: [
+				{
+					type: 'dropdown',
+					label: 'zone/up/down',
+					id: 'volume',
+					default: 'volume_up',
+					choices: [
+						{ label: 'Main Zone volume up', id: 'MVUP' },
+						{ label: 'Main Zone volume down', id: 'MVDOWN' },
+						{ label: 'Zone 2 volume up', id: 'Z2UP' },
+						{ label: 'Zone 2 volume down', id: 'Z2DOWN' },
+						{ label: 'Zone 3 volume up', id: 'Z3UP' },
+						{ label: 'Zone 3 volume down', id: 'Z3DOWN' },
+					],
+				},
+			],
+			callback: async (event) => {
+				var cmd = event.options.volume + '\r'
+				console.log('command sent to denon ->', cmd)
+				self.socket.send(cmd)
+			},
+		},
+		set_volume: {
+			name: 'Set Volume',
+			options: [
+				{
+					type: 'dropdown',
+					label: 'select zone',
+					id: 'zone',
+					default: 'MV',
+					choices: [
+						{ label: 'Main Zone', id: 'MV' },
+						{ label: 'Zone 2', id: 'Z2' },
+						{ label: 'Zone 3', id: 'Z3' },
+					],
+				},
+				{
+					type: 'dropdown',
+					id: 'set_volume',
+					default: '00',
+					choices: CHOICES_VOLUME,
+				},
+			],
+			callback: async (event) => {
+				var cmd = event.options.zone + String(event.options.set_volume).padStart(2, '0') + '\r'
```

Comment:
this one seems to be throwing an error -- 
`23.04.08 13:53:38 Instance/Wrapper/denon: Error executing action: [event.options.set_volume].padstart is not a function`
I think 'padStart' may only be a part of the String() function?
anyhow, since I'm forcing the string padding when creating the options, its not necessary inside the action callback
this seems to be working as expected -->  `var cmd = `${event.options.zone}${event.options.set_volume}\r``